### PR TITLE
Containerfactory SPI

### PR DESCRIPTION
**Pull request by tysonnorris**
This SPI interface allows alternate container management strategies, e.g. docker, mesos, kubernetes, etc. 

Default impl is DockerContainerFactoryProvider, which has same behavior as before.

### CODE REVIEW

File: core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainerFactory.scala [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+class DockerContainerFactoryProvider extends ContainerFactoryProvider {
+    override def getContainerFactory(actorSystem: ActorSystem, logging: Logging, config: WhiskConfig): ContainerFactory =
+        new DockerContainerFactory()(actorSystem.dispatcher, logging)
+}
+
+object DockerContainerFactoryProvider extends SpiFactory[ContainerFactoryProvider] {
+    override def apply(dependencies: Dependencies): ContainerFactoryProvider = new DockerContainerFactoryProvider()
+}
```
Comment (markusthoemmes) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainerFactory.scala]: I do still think that these SPI implementations are too verbose, the intermediary steps (including `SingletonSpiFactory`) seems unnecessary bloat. Shall we refactor those bits before getting a lot of interfaces in, that we want to change soon anyway?

File: common/scala/src/main/scala/whisk/core/containerpool/Container.scala [hunk 2]
```diff
@@ -66,15 +74,3 @@ case class BlackboxStartupError(msg: String) extends ContainerStartupError(msg)
...
-    def duration = Duration.create(end.toEpochMilli() - start.toEpochMilli(), MILLISECONDS)
-}
-
-object Interval {
-    /** An interval starting now with zero duration. */
-    def zero = {
-        val now = Instant.now
-        Interval(now, now)
-    }
-}
```
Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/core/containerpool/Container.scala]: Why did you move these bits?

File: core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainerFactory.scala (hunk 1 above)
Comment (tysonnorris) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainerFactory.scala]: What specifically do you have in mind? Are you sure this can be simplified further?

File: core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala [hunk 3]
```diff
@@ -120,7 +122,7 @@ object DockerContainer {
  * @param id the id of the container
  * @param ip the ip of the container
  */
-class DockerContainer(id: ContainerId, ip: ContainerIp)(
+class DockerContainer(val id: ContainerId, val ip: ContainerIp)(
```
Comment (markusthoemmes) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala]: Do these need to be public?

File: common/scala/src/main/scala/whisk/core/containerpool/Container.scala (hunk 2 above)
Comment (tysonnorris) [inline common/scala/src/main/scala/whisk/core/containerpool/Container.scala]: I'm suggesting that container management can be handled outside of an invoker, therefore Container related interfaces do not belong there.

Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/core/containerpool/Container.scala]: Right, that's fine. But why move these classes out of `Container.scala`?

Comment (tysonnorris) [reply common/scala/src/main/scala/whisk/core/containerpool/Container.scala]: I'm not sure why I moved them originally - moved back to Container.scala

File: core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala (hunk 3 above)
Comment (tysonnorris) [reply core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala]: id/ip are needed by the HTTP API during initialize() and run() - latest changes have these as protected val, but also has the HTTP API methods initialize() and run() moved into a separate class so they can be shared - see ContainerApi.scala

File: core/invoker/src/main/scala/whisk/core/invoker/InvokerReactive.scala [hunk 4]
```diff
@@ -72,31 +69,12 @@ class InvokerReactive(config: WhiskConfig, instance: InstanceId, producer: Messa
...
-    }
-    cleanup()
-    sys.addShutdownHook(cleanup())
 
     /** Factory used by the ContainerProxy to physically create a new container. */
-    val containerFactory = SpiLoader.get[ContainerFactoryProvider]().getContainerFactory(actorSystem, logging, config).createContainer _
+    val containerFactory = SpiLoader.get[ContainerFactoryProvider].getContainerFactory(actorSystem, logging, config, instance)
+    val containerFactoryFunction = containerFactory.createContainer _
+    containerFactory.cleanup()
+    sys.addShutdownHook(containerFactory.cleanup())
```
Comment (jcrossley3) [inline core/invoker/src/main/scala/whisk/core/invoker/InvokerReactive.scala]: I think this just effectively runs `cleanup` twice at startup. Shouldn't it be `sys.addShutdownHook { containerFactory.cleanup() }` instead?

Comment (tysonnorris) [reply core/invoker/src/main/scala/whisk/core/invoker/InvokerReactive.scala]: This is the same syntax that was there before; I tested it and it does the right thing - runs once at startup, and once at shutdown.

Comment (markusthoemmes) [reply core/invoker/src/main/scala/whisk/core/invoker/InvokerReactive.scala]: For posterity:

```scala
def addShutdownHook(body: => Unit): ShutdownHookThread = ShutdownHookThread(body)
```

`body` is a by-name argument, which is only evaluated once its used. I agree it looks like the function is actually called, but braces or parentheses do no make a difference in Scala here.

Comment (jcrossley3) [reply core/invoker/src/main/scala/whisk/core/invoker/InvokerReactive.scala]: Thanks for the explanation!

File: core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainerFactory.scala (hunk 1 above)
Comment (tysonnorris) [reply core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainerFactory.scala]: Simplified variant implemented now (no more SpiFactory)

File: common/scala/src/main/scala/whisk/core/containerpool/ContainerApi.scala [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+import whisk.common.Logging
+import whisk.common.LoggingMarkers
+import whisk.common.TransactionId
+import whisk.core.entity.ActivationResponse
+import whisk.core.entity.ActivationResponse.ContainerConnectionError
+import whisk.core.entity.ActivationResponse.ContainerResponse
+import whisk.core.entity.size._
+import whisk.http.Messages
+
+trait ContainerApi { container: Container =>
```
Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/core/containerpool/ContainerApi.scala]: Is this type narrowing useful? I thought about doing the same abstraction a while ago aswell. Would it make sense to instead model it as 

```scala
abstract class HttpContainer(id: ContainerId, ip: ContainerIp)(implicit ec: ExecutionContext, logging: Logging) extends Container
```

File: core/invoker/src/main/scala/whisk/core/invoker/InvokerReactive.scala [hunk 6]
```diff
@@ -81,53 +75,14 @@ class InvokerReactive(config: WhiskConfig, instance: InstanceId, producer: Messa
...
-        cpuShares = config.invokerCoreShare.toInt,
-        environment = Map("__OW_API_HOST" -> config.wskApiHost),
-        network = config.invokerContainerNetwork,
-        dnsServers = config.invokerContainerDns,
-        name = Some(name))
-    }
+    SpiLoader
+      .get[ContainerFactoryProvider]
+      .getContainerFactory(actorSystem, logging, config, instance)
+  val containerFactoryFunction = containerFactory.createContainer _
```
Comment (markusthoemmes) [inline core/invoker/src/main/scala/whisk/core/invoker/InvokerReactive.scala]: No need to create an additional `val` I think, just put `containerFactory.createContainer _` into `props` directly.

File: common/scala/src/main/scala/whisk/core/containerpool/ContainerApi.scala [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    }
+    Future {
+      http.post(path, body, retry)
+    }.map { response =>
+      val finished = Instant.now()
+      RunResult(Interval(started, finished), response)
+    }
+  }
+
+  def destroyApi()(implicit transid: TransactionId): Future[Unit] = {
```
Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/core/containerpool/ContainerApi.scala]: Why `destroyApi` vs. just `destroy`?

File: common/scala/src/main/scala/whisk/core/containerpool/ContainerApi.scala (hunk 5 above)
Comment (markusthoemmes) [reply common/scala/src/main/scala/whisk/core/containerpool/ContainerApi.scala]: Oh got it, this is the "formalized" version of something the runtime needs to implement and stuff. Hmm. Might be sensible. Thinking about it.

File: common/scala/src/main/scala/whisk/core/containerpool/ContainerFactory.scala [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    val dockerRunParameters = new HashMap[String, Set[String]] with MultiMap[String, String]
+    dockerRunParameters.addBinding("--cap-drop", "NET_RAW")
+    dockerRunParameters.addBinding("--cap-drop", "NET_ADMIN")
+    dockerRunParameters.addBinding("--ulimit", "nofile=1024:1024")
+    dockerRunParameters.addBinding("--pids-limit", "1024")
+    val dnsServers = config.invokerContainerDns
+    dnsServers.map(d => dockerRunParameters.addBinding("--dns", d))
+    dockerRunParameters
+  }
+}
```
Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/core/containerpool/ContainerFactory.scala]: Not sure if using a `Map` here is useful. For additional arguments I found it mostly useful to just use a flat `Seq` and pass stuff that way.

Alternatively we could have a type encapsulating all the options on a docker run and propagate it that way. But please do not use a mutable map here. If you think a Map is the right thing to do, let's at least use an immutable one (I know you can't use MultiMap then)

File: common/scala/src/main/scala/whisk/core/containerpool/ContainerApi.scala (hunk 5 above)
Comment (tysonnorris) [reply common/scala/src/main/scala/whisk/core/containerpool/ContainerApi.scala]: e.g. `MesosTask extends Container with ContainerApi` - an alternative is to move the HTTP contract definition into Container?

File: common/scala/src/main/scala/whisk/core/containerpool/ContainerApi.scala (hunk 7 above)
Comment (tysonnorris) [reply common/scala/src/main/scala/whisk/core/containerpool/ContainerApi.scala]: I wasn't really sure how best to call `ContainerApi.destroy()` from `DockerContainer.destroy()`, since DockerContainer is now defined `with ContainerApi`. 
Ideas?

File: common/scala/src/main/scala/whisk/core/containerpool/ContainerFactory.scala (hunk 8 above)
Comment (tysonnorris) [reply common/scala/src/main/scala/whisk/core/containerpool/ContainerFactory.scala]: At least partly I used Map because a) all the options I think that should be settable here are key-value and b) I happen to know in mesos we will have to translate this into a Map :(

I think the options will boil down to either using Seq (which at least some alternative container factories will have to decode to a Map) or writing an immutable version of MultiMap. I'm not wild about either of these prospects, let me know if you have other ideas.

File: common/scala/src/main/scala/whisk/core/containerpool/ContainerApi.scala (hunk 5 above)
Comment (markusthoemmes) [reply common/scala/src/main/scala/whisk/core/containerpool/ContainerApi.scala]: Right, we could just fold it into `Container` itself or have an intermediary type between `HttpContainer` which is exactly about the Http communication. Both sound sensible to me, cause I think all the container impls will just have `with ContainerApi`.

File: common/scala/src/main/scala/whisk/core/containerpool/ContainerApi.scala (hunk 7 above)
Comment (markusthoemmes) [reply common/scala/src/main/scala/whisk/core/containerpool/ContainerApi.scala]: Yeah I guess that's why I'd prefer it folded into either Container or have HttpContainer as an extension: Implementations can delegate to `super.destroy()`, which makes the impl. more transparent maybe?

File: common/scala/src/main/scala/whisk/core/containerpool/ContainerFactory.scala (hunk 8 above)
Comment (markusthoemmes) [reply common/scala/src/main/scala/whisk/core/containerpool/ContainerFactory.scala]: As seen in the method signature, `MultiMap` is just some sugar over `Map[String, Set[String]]`, so you can translate your mutable code to:

```scala
Map(
  "--cap-drop" -> Set("NET_RAW", "NET_ADMIN"),
  "--ulimit" -> Set("nofile=1024:1024"),
  "--pids-limit" -> Set("1024"),
  "--dns" -> config.invokerContainerDns.toSet
)
```

Which I'd prefer to keep the interface immutable. MultiMap-sugar is arguably not helping much here anyway?

File: core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala [hunk 9]
```diff
@@ -66,27 +59,25 @@ object DockerContainer {
...
+    val environmentArgs = environment.flatMap {
       case (key, value) => Seq("-e", s"$key=$value")
-    }.flatten
+    }
 
-    val dnsArgs = dnsServers.map(Seq("--dns", _)).flatten
+    val params = dockerRunParameters.flatMap {
+      case (key, valueList) =>
+        valueList.map {
+          case (value) => Seq(key, value)
```
Comment (markusthoemmes) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala]: Drop the unnecessary parentheses around `value`.

File: common/scala/src/main/scala/whisk/core/containerpool/ContainerFactory.scala (hunk 8 above)
Comment (markusthoemmes) [reply common/scala/src/main/scala/whisk/core/containerpool/ContainerFactory.scala]: Also: Shouldn't this method exist only for the docker implementation?

File: core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala [hunk 10]
```diff
@@ -66,27 +59,25 @@ object DockerContainer {
...
-    val environmentArgs = environment.map {
+    val environmentArgs = environment.flatMap {
       case (key, value) => Seq("-e", s"$key=$value")
-    }.flatten
+    }
 
-    val dnsArgs = dnsServers.map(Seq("--dns", _)).flatten
+    val params = dockerRunParameters.flatMap {
+      case (key, valueList) =>
+        valueList.map {
```
Comment (markusthoemmes) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala]: Could be `flatMap` to not iterate again.

File: core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainerFactory.scala [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+      cpuShares = config.invokerCoreShare.toInt,
+      environment = Map("__OW_API_HOST" -> config.wskApiHost),
+      network = config.invokerContainerNetwork,
+      dnsServers = config.invokerContainerDns,
+      name = Some(name),
+      parameters)
+  }
+
+  /** Perform cleanup on init */
+  override def init: Unit = cleanup()
```
Comment (markusthoemmes) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainerFactory.scala]: Add parentheses to `init` and `cleanup`s definition please. Both are sideeffecting and not "get like" behavior with no-sideeffects.

File: common/scala/src/main/scala/whisk/core/containerpool/ContainerFactory.scala (hunk 8 above)
Comment (tysonnorris) [reply common/scala/src/main/scala/whisk/core/containerpool/ContainerFactory.scala]: Agree on immutable MultiMap, will make those changes.

Run parameters are possible in other non-docker-cli container management systems (e.g. mesos, guessing kube as well), so it should be useable by any SPI impl.

File: core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala (hunk 10 above)
Comment (tysonnorris) [reply core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala]: FlatMap on valueList (a Set) caused the duplicate keys to be removed - I changed it to `valueList.toList.flatMap` to avoid that, but I'm not sure that is what you're looking for, performance-wise. Let me know if you have suggestions.

File: common/scala/src/main/scala/whisk/core/containerpool/ContainerFactory.scala (hunk 8 above)
Comment (markusthoemmes) [reply common/scala/src/main/scala/whisk/core/containerpool/ContainerFactory.scala]: Still the method is called `dockerRunParameters` but it's part of the general `ContainerFactory`. It's not part of any interface, so it should live in `DockerContainerFactory`?

Comment (tysonnorris) [reply common/scala/src/main/scala/whisk/core/containerpool/ContainerFactory.scala]: Are you suggesting to rename it to `containerRunParameters`? or that there may be some non-docker container support, like lxc/oci/etc? 

The problem with "moving it to DockerContainerFactory" is that DockerContainerFactory shouldn't be used if some other container factory is used, but these parameters are still important. 

This may be confusing terminology, but currently the notion of "container" exists only in the form of "docker container", but the ContainerFactory is allowing multiple approaches to creating "docker containers" (as opposed to creating "other non-docker containers"). Its a difference of using docker cli to create a docker container, vs using some other system (which in turn uses docker cli) to create a docker container.

**Comment by markusthoemmes** [comment]
PG1 2102 🔵

**Comment by markusthoemmes** [comment]
@tysonnorris this is good to be merged but I wanted to squash it locally which failed miserably because of the merge commits inside. I got conflicts all about. Can you please rebase this on latest master and squash it to one commit with an appropriate title?

**Comment by tysonnorris** [comment]
@markusthoemmes I did it this way, since I don't relish reapplying every merge conflict:
```
git reset --soft master
git commit -m "ContainerFactory SPI"
git push origin +containerfactory
```
Let me know if you think that is problematic?

**Comment by markusthoemmes** [comment]
Great job, thanks for doing this 🎉
